simple_metering.cpp

```diff
@@ -0,0 +1,200 @@
+                    else if (interfaceMode == 0x0103) { mode = 7; }     // P1 Dutch Standard – DSMR 2.3 Version
+                    else if (interfaceMode == 0x0104) { mode = 8; }     // P1 Dutch Standard – DSMR 4.0 Version
+                    
+                    item = sensor->item(RConfigInterfaceMode);
+                    if (item && item->toNumber() != mode && mode > 0 && mode < 9)
+                    {
+                        item->setValue(mode);
+                        enqueueEvent(Event(RSensors, RConfigInterfaceMode, sensor->id(), item));
+                        configUpdated = true;
+                    }
+                }
+                sensor->setZclValue(updateType, ind.srcEndpoint(), METERING_CLUSTER_ID, attrId, attr.numericValue());
+            }
+                break;
+
+            default:
+                break;
+            }
+        }
+
+        if (stateUpdated)
+        {
+            sensor->updateStateTimestamp();
+            enqueueEvent(Event(RSensors, RStateLastUpdated, sensor->id()));
+        }
+
+        if (configUpdated || stateUpdated)
+        {
+            updateSensorEtag(&*sensor);
+            sensor->setNeedSaveDatabase(true);
+            queSaveDb(DB_SENSORS, DB_SHORT_SAVE_DELAY);
+        }
+    }
+}
+
+/*! Write Attribute on thermostat cluster.
+ *  Iterate over every day and get schedule for each day.
+   \param task - the task item
+   \param attrId
+   \param attrType
+   \param attrValue
+   \return true - on success
+           false - on error
+ */
+bool DeRestPluginPrivate::addTaskSimpleMeteringReadWriteAttribute(TaskItem &task, uint8_t readOrWriteCmd, uint16_t attrId, uint8_t attrType, uint32_t attrValue, uint16_t mfrCode)
+{
+    if (readOrWriteCmd != deCONZ::ZclReadAttributesId && readOrWriteCmd != deCONZ::ZclWriteAttributesId)
+    {
+        DBG_Printf(DBG_INFO, "Invalid command for simple metering cluster %d\n", readOrWriteCmd);
+        return false;
+    }
+
+    task.taskType = TaskSimpleMetering;
+
+    task.req.setClusterId(METERING_CLUSTER_ID);
+    task.req.setProfileId(HA_PROFILE_ID);
+
+    task.zclFrame.payload().clear();
+    task.zclFrame.setSequenceNumber(zclSeq++);
+    task.zclFrame.setCommandId(readOrWriteCmd);
+    task.zclFrame.setFrameControl(deCONZ::ZclFCProfileCommand |
+            deCONZ::ZclFCDirectionClientToServer |
+            deCONZ::ZclFCDisableDefaultResponse);
+
+    if (mfrCode != 0x0000)
+    {
+        task.zclFrame.setFrameControl(task.zclFrame.frameControl() | deCONZ::ZclFCManufacturerSpecific);
+        task.zclFrame.setManufacturerCode(mfrCode);
+    }
+
+    // payload
+    QDataStream stream(&task.zclFrame.payload(), QIODevice::WriteOnly);
+    stream.setByteOrder(QDataStream::LittleEndian);
+
+    if (readOrWriteCmd == deCONZ::ZclWriteAttributesId)
+    {
+        stream << static_cast<quint16>(attrId);
+        stream << static_cast<quint8>(attrType);
```

Comment:
Doesn't need casting as parameters already have the right type.

Reply:
kk